packages/nouns-contracts/contracts/governance/NounsDAOLogicV3.sol

```diff
@@ -419,6 +424,59 @@ contract NounsDAOLogicV3 is NounsDAOStorageV2, NounsDAOEventsV2 {
         emit ProposalVetoed(proposalId);
     }
 
+    function ragequit(uint256[] calldata nounIds) external {
+        uint256 supply = nounsCirculatingSupply();
+
+        // transfer all nouns to timelock
+        // will revert if not owner or has approval for all nouns
+        // TODO: should this be allowed only for owner of the nouns?
+        for (uint256 i = 0; i < nounIds.length; ++i) {
+            nouns.transferFrom(msg.sender, address(timelock), nounIds[i]);
+        }
+
+        ////// ETH redeem
+
+        // calculate ETH pro rata share
+        uint256 proRataShare = (nounIds.length * address(timelock).balance) / supply;
+
+        // apply penalty
+        uint256 amountOfEthToRedeem = ((10_000 - ragequitPenaltyBPs) * proRataShare) / 10_000;
+
+        // send ETH to msg.sender
+        timelock.redeem(msg.sender, amountOfEthToRedeem);
+
+        ////// ERC20s redeem
+        for (uint256 i = 0; i < redeemableAssets.length; ++i) {
+            address asset = redeemableAssets[i];
+
+            proRataShare = (nounIds.length * IERC20(asset).balanceOf(address(timelock))) / supply;
+            uint256 amountOfTokensToRedeem = ((10_000 - ragequitPenaltyBPs) * proRataShare) / 10_000;
+            IERC20(asset).transferFrom(address(timelock), msg.sender, amountOfTokensToRedeem);
```

Comment:
we want to move this call into a new timelock function to remove the need for token approvals (in line with the other comment, letting minorities quit with ERC20 shares without needing majority permission)